### ISSUE TITLE
Redcarpet plugin not handling Guesser::Ambiguous

### ISSUE THREAD

**Issue by johnfairh** [rouge-ruby/rouge#1343]
Looking for some advice on how best to fix -- discussion after the form.

**Describe the bug**
If the Rouge Redcarpet plugin tries to guess the code language and the guess is ambiguous then it propagates an exception back up the stack which abandons the entire document.

**To Reproduce**
1. Set up Redcarpet with Rouge plugin.
2. Pass markdown through Redcarpet:
<pre>
```
< !DOCTYPE math>
```
</pre>
(without the space between < and !)

To see this particular ambiguity without redcarpet:
```shell
echo '<!DOCTYPE math>' | rougify highlight
```

**Expected behavior**
Fall back to either plaintext lexer or one of the ambiguees.

**System**
Any, though only seen in practice since Mason lexer added in Rouge 3.7.

-----
The redcarpet plugin is [very simple](https://github.com/rouge-ruby/rouge/blob/d31841900ff675661d124e82f8b2d854fe11cb87/lib/rouge/plugins/redcarpet.rb#L12) -- `code` is the stuff from within the triple backticks and `language` is the optional language identifier.

So this just calls [`Lexer.find_fancy`](https://github.com/rouge-ruby/rouge/blob/d31841900ff675661d124e82f8b2d854fe11cb87/lib/rouge/lexer.rb#L55).  In the case when [no language is supplied by the user](https://github.com/rouge-ruby/rouge/blob/d31841900ff675661d124e82f8b2d854fe11cb87/lib/rouge/lexer.rb#L79) we go to [`Lexer.guess`](https://github.com/rouge-ruby/rouge/blob/d31841900ff675661d124e82f8b2d854fe11cb87/lib/rouge/lexer.rb#L169) who raises an exception if there is ambiguity.

The particular ambiguity with DOCTYPE has happened because both the [xml](https://github.com/rouge-ruby/rouge/blob/d31841900ff675661d124e82f8b2d854fe11cb87/lib/rouge/lexers/xml.rb#L18) and [mason](https://github.com/rouge-ruby/rouge/blob/d31841900ff675661d124e82f8b2d854fe11cb87/lib/rouge/lexers/mason.rb#L20) lexers respond to 'detect' for non-xml/html doctypes.

My first reactive fix is to delete `Lexers::Mason.detect?` -- I don't think it can be right for this dialect to claim all unclaimed doctypes?  But I barely understand what Mason is.

Then a wider question: do users of `find_fancy` expect it to raise exceptions?  It's not documented as such and none of the Rouge-internal users seem to be coded that way.  So maybe we should change that interface to return one of the ambiguous lexers.  This is a very old public interface though -- probably better to update the docs and update its callers to handle the exception and use one of the lexers therein?

**Comment by pyrmont** [comment]
@miparnisari It seems a mistake to have the Mason lexer match against all documents with a doctype (excepting HTML and XML doctypes). I'm strongly inclined to remove this. Do you remember why it was done this way?

**Comment by pyrmont** [comment]
@johnfairh Mason is a Perl templating language. I'm not sure why it matches against almost all doctypes; that definitely seems like a mistake. @miparnisari did the hard work of adding Mason to Rouge (and I know she uses the lexer) so will check in with her :)

**Comment by miparnisari** [comment]
@pyrmont hmm no I can't remember. Feel free to remove it,

**Comment by johnfairh** [comment]
Ok, will remove that and add exception handling.

**Comment by pyrmont** [comment]
This has been fixed by #1348 and #1349. Thanks @johnfairh!
